src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -364,38 +365,38 @@ private function getTagVersions(array $tagsByKey, bool $persistTags): array
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
             $knownTagVersion = $this->knownTagVersions[$tag] ?? [0, null];
-            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now - $knownTagVersion[0] >= $this->knownTagVersionsTtl) {
-                // reuse previously fetched tag versions up to the ttl
+            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now > $knownTagVersion[0]) {
+                // reuse previously fetched tag versions until the expiration
+                unset($this->knownTagVersions[$tag];
                 $fetchTagVersions = true;
             }
-            unset($this->knownTagVersions[$tag]); // For LRU tracking
-            if ([0, null] !== $knownTagVersion) {
-                $this->knownTagVersions[$tag] = $knownTagVersion;
-            }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        if (\count($this->knownTagVersions) > self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS && $now > reset($this->knownTagVersions)[0]) {
```

Comment:
Do we need the `$now > reset(...)` part? Why is it needed?

Reply:
This is needed to check if there are expired tags in the cache. The first tag is the oldest one.

Reply:
I get that, but doesn't this added check break what MAX_NUMBER_OF_KNOWN_TAG_VERSIONS means?

Reply:
I agree that we get soft limit but we don't need it to be hard. We need some mechanism to purge needless tags which may stay in cache forever. MAX_NUMBER_OF_KNOWN_TAG_VERSIONS donotes acceptable size of the cache.

Reply:
MAX_NUMBER_OF_KNOWN_TAG_VERSIONS means ALLOWED_NUMBER_OF_EXPIRED_KNOWN_TAG_VERSIONS
However, I doubt that it ought to be renamed this way. Your thoughts?

Reply:
I don't see the benefit of the reset check so I'd drop it :)

Reply:
I wouldn't drop it because then we'll get hard limit which cannot be controlled by programmers. 

This means that, for example, big sets of tags won't fit in the cache and will be cut off pretty often, and that TTL won't work as expected. With hard limit for size, `$knownTagVersionTtl` becomes `$knownTagVersionMaxTtl`. You cannot effectively control the TTL anymore and lose the possibility to improve performance by extending it.

Reply:
My idea was to purge only expired tags.
```suggestion
        if (\count($this->knownTagVersions) > self::MAX_NUMBER_OF_EXPIRED_TAG_VERSIONS && $now > reset(array_slice($this->knownTagVersions, self::MAX_NUMBER_OF_EXPIRED_TAG_VERSIONS, 1))[0]) {
            array_splice($this->knownTagVersions, 0, self::MAX_NUMBER_OF_EXPIRED_TAG_VERSIONS);
        }
```
I simplified this logic a bit.

Reply:
Then, what about removing any limit, and just remove expired items? Because the current logic doesn't enforce the limit anyway, and it can also remove non-expired items :shrug: 

```php
while ($now > (current($this->knownTagVersions)[0] ?? INF)) {
    array_shift($this->knownTagVersions);
}
```

Reply:
(snippet updated)